clio-client/src/main/scala/org/broadinstitute/clio/client/ClioClient.scala

```diff
@@ -1,47 +1,69 @@
 package org.broadinstitute.clio.client
 
-import org.broadinstitute.clio.client.commands.CommandDispatch
-import org.broadinstitute.clio.client.parser.{BaseArgs, BaseParser}
-import org.broadinstitute.clio.client.webclient.ClioWebClient
 import akka.actor.ActorSystem
 import akka.http.scaladsl.model.HttpResponse
+import akka.http.scaladsl.model.headers.OAuth2BearerToken
+import caseapp._
+import com.typesafe.scalalogging.LazyLogging
+import org.broadinstitute.clio.client.commands.{CommandDispatch, CommandType}
+import org.broadinstitute.clio.client.commands.Commands._
 import org.broadinstitute.clio.client.util.IoUtil
+import org.broadinstitute.clio.client.webclient.ClioWebClient
 
 import scala.concurrent.{ExecutionContext, Future}
 import scala.util.{Failure, Success}
 
-object ClioClient extends App {
+object ClioClient
+    extends CommandAppWithPreCommand[CommonOptions, CommandType]
+    with LazyLogging {
+  override val progName = "clio-client"
+  override val appName = "Clio Client"
+
+  override val appVersion: String = ClioClientConfig.Version.value
+
+  //not sure this is the best way to pass common opts but case-app docs are spotty at best
+  var commonOptions: CommonOptions = _
+
+  override def beforeCommand(options: CommonOptions,
+                             remainingArgs: Seq[String]): Unit = {
+    commonOptions = options
+    if (remainingArgs.nonEmpty) {
+      logger.error(s"Found extra arguments: ${remainingArgs.mkString(" ")}")
+      sys.exit(1)
+    }
+  }
+
   implicit val system: ActorSystem = ActorSystem("clio-client")
   import system.dispatcher
   sys.addShutdownHook({ val _ = system.terminate() })
 
-  val webClient = new ClioWebClient(
+  val webClient: ClioWebClient = new ClioWebClient(
     ClioClientConfig.ClioServer.clioServerHostName,
     ClioClientConfig.ClioServer.clioServerPort,
     ClioClientConfig.ClioServer.clioServerUseHttps
   )
-  val commandDispatch = new CommandDispatch(webClient, IoUtil)
 
-  val client = new ClioClient(commandDispatch)
+  override def run(command: CommandType, remainingArgs: RemainingArgs): Unit = {
+    implicit val bearerToken: OAuth2BearerToken = commonOptions.bearerToken
+
+    val commandDispatch = new CommandDispatch(webClient, IoUtil)
 
-  client.execute(args) onComplete {
-    case Success(_) => System.exit(0)
-    case Failure(_) => System.exit(1)
+    val client = new ClioClient(commandDispatch)
+
+    client.execute(command) onComplete {
+      case Success(_) => sys.exit(0)
+      case Failure(_) => sys.exit(1)
+    }
   }
 }
 
 class ClioClient(commandDispatch: CommandDispatch) {
```

Comment:
Not much point to this class now.

Reply:
aside from giving tests a call execute without working about sys.exit I agree. I'll try to factor out the tests that call execute directly.